### ISSUE TITLE
It would be great for Wechaty to support experimental features in puppet

### ISSUE THREAD

**Issue by hcfw007** [wechaty/wechaty#2316]
**Is your feature request related to a problem? Please describe.**
Recently we have been working on location messages and moments. I do understand that Wechaty has such support or is working on it, but in the future, if Wechaty can provide a method to call any function directly from puppet-implementation, that would be great.

**Describe the solution you'd like**
something like
```
class Wechaty {
// ... other code

  directPuppetCall(func, ...args) {
    return this.puppet[func](args)
  }
}
```
**Describe alternatives you've considered**
Or provide a public reference of puppet-implementation instance in Wechaty class?

**Additional context**
may require additional changes in wechaty-puppet and puppet-service

[enhancement]

**Comment by huan** [comment]
We have had this discussion a long time ago and I'm very happy that there's an issue with discussing it.

I believe, instead of adding this to the Wechaty class, it will be very easy to be achieved by a simple function:

```ts
const directPuppetCallWechaty<P = any, R=any> = (wechaty: Wechaty) => (func: string) => (...args: P): R {
  return wechaty.puppet[func](args)
}

const puppetDirectFunc = directPuppetCallWechaty(wechaty)('directFunc')
const ret = await puppetDirectFunc()
```

So there's no need to add anything to Wechaty at all.

Please feel free to let me know what you think about the design.

**Comment by hcfw007** [comment]
Yes, since puppet is a public property of Wechaty, we surely can do that.
The reasons I believe we should add that in Wechaty are:
1. We can add some public logic, such as ```log.warn('You are calling unsupported method')```
2. We can add that to puppet-service. So far it will throw an Error.

**Comment by hcfw007** [comment]
I just found a scenario for this proposal.
Say, we have a Lark puppet. Lark supports 'task' module, so that if we can create a task when receiving certain messages, that would be great. However most IMs does not have such module, so it's not necessary for Wechaty to support it. In this case, the developer of Lark puppet can add this function like a 'private feature'. In this case, puppet-service support of directPuppetCall might be a good solution.

**Comment by hcfw007** [comment]
So far we use a walk-around with ding-dong to achieve this. With the same spirit, I'd also suggest that wechaty could support some additional data for almost all payloads. Different IMs or even same IM but different puppet could provide different information for the same Contact, Room or Message. This kind of messages are usually just plain string and No need to create a specific field to do so.
e.g.: 
current contact interface:
```ts
export interface ContactPayload {
  id     : string,
  gender : ContactGender,
  type   : ContactType,
  // ... some other info
  weixin?    : string,
  // ... some other info
  coworker?    : boolean,
}
```
should change to 
```ts
export interface ContactPayload {
  id     : string,
  gender : ContactGender,
  type   : ContactType,
  // ... some other info
  additionalInfo?: string // could contain wexin and coworker
}```
The reason I put wexin and coworker into additionalInfo is this kind of information all not universal. Coworker is only for wecom, and people might have many IM contact info instead of just weixin, we cannot put everything in it. Otherwise why don't we also add twittwer, facebook and linkedIn?
Also this will allow puppet to pass some IM or puppet specific info to the user and make wechaty much more flexible.

**Comment by hcfw007** [comment]
Maybe this should be moved to a new thread.
#2329

**Comment by hcfw007** [comment]
Issue should be closed since discusses on this matter is over.
